### PR TITLE
feat(migrate diff): add migrate diff CLI command

### PR DESCRIPTION
**Pull request by Jolg42**
Related: https://github.com/prisma/migrations-team/issues/313

Notes:
Design doc on notion: https://www.notion.so/prismaio/Design-doc-first-batch-of-migrate-primitives-48be767a7adf4c23b3116d1347f218c8

### CODE REVIEW

**Review by tomhoule — APPROVED** [review]
lgtm :+1:

Nice that the places we need to improve are marked and tested.

File: packages/migrate/src/commands/MigrateDiff.ts [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+      Number(Boolean(args['--to-schema-datasource'])) +
+      Number(Boolean(args['--to-schema-datamodel'])) +
+      Number(Boolean(args['--to-url'])) +
+      Number(Boolean(args['--to-migrations']))
+
+    // One of --to or --from is required
+    if (numberOfFromParamaterProvided !== 1 || numberOfToParamaterProvided !== 1) {
+      const errorMessages: string[] = []
+      if (numberOfFromParamaterProvided !== 1) {
+        errorMessages.push(`${numberOfFromParamaterProvided} \`--from-...\` paramater(s) provided. 1 must be provided.`)
```
Comment (millsp) [inline packages/migrate/src/commands/MigrateDiff.ts:165]:
```suggestion
        errorMessages.push(`${numberOfFromParamaterProvided} \`--from-...\` parameter(s) provided. 1 must be provided.`)
```

File: packages/migrate/src/commands/MigrateDiff.ts [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+      }
+      if (numberOfToParamaterProvided !== 1) {
+        errorMessages.push(`${numberOfToParamaterProvided} \`--to-...\` paramater(s) provided. 1 must be provided.`)
```
Comment (millsp) [inline packages/migrate/src/commands/MigrateDiff.ts:168]:
```suggestion
        errorMessages.push(`${numberOfToParamaterProvided} \`--to-...\` parameter(s) provided. 1 must be provided.`)
```

Comment (millsp) [reply packages/migrate/src/commands/MigrateDiff.ts:168]: Variable can be renamed too.

File: packages/migrate/src/__tests__/MigrateDiff.test.ts [hunk 3]
```diff
@@ -0,0 +1,513 @@
...
+        await MigrateDiff.new().parse([
+          '--preview-feature',
+          '--from-schema-datasource=./doesnoexists.prisma',
+          '--to-empty',
+        ])
+      } catch (e) {
+        expect(e.code).toEqual(undefined)
+        expect(e.message).toMatchInlineSnapshot(`
+          Could not find a schema.prisma file that is required for this command.
+          You can either provide it with --schema, set it as \`prisma.schema\` in your package.json or put it into the default location ./prisma/schema.prisma https://pris.ly/d/prisma-schema-location
```
Comment (millsp) [inline packages/migrate/src/__tests__/MigrateDiff.test.ts:94]: Does this work with the client as well?

File: packages/migrate/src/__tests__/MigrateDiff.test.ts [hunk 4]
```diff
@@ -0,0 +1,513 @@
...
+  })
+
+  describe('sqlite', () => {
+    // TODO next 2 tests: is it expected to not fail when diffing from/to an unexisting sqlite db?
+    it('should diff --from-empty --to-url=file:doesnotexists.db', async () => {
+      ctx.fixture('schema-only-sqlite')
+
+      const result = MigrateDiff.new().parse(['--preview-feature', '--from-empty', '--to-url=file:doesnotexists.db'])
+      await expect(result).resolves.toMatchInlineSnapshot(``)
+      expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
```
Comment (millsp) [inline packages/migrate/src/__tests__/MigrateDiff.test.ts:122]: Shouldn't this produce a diff too?

Comment (millsp) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:122]: Or does it produce a diff only when there's something to compare on both ends?

**Review by millsp — CHANGES_REQUESTED** [review]
Solid testing, nice work :100: :clap:

File: packages/migrate/src/commands/MigrateDiff.ts (hunk 1 above)
Comment (Jolg42) [reply packages/migrate/src/commands/MigrateDiff.ts:165]: Nice one 😆

File: packages/migrate/src/commands/MigrateDiff.ts (hunk 2 above)
Comment (Jolg42) [reply packages/migrate/src/commands/MigrateDiff.ts]: All done in https://github.com/prisma/prisma/pull/11440/commits/2780e8290d6f4102ae76a1926699858d2e1f2b68

Thanks!

File: packages/migrate/src/__tests__/MigrateDiff.test.ts (hunk 4 above)
Comment (Jolg42) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:122]: Here it's empty to empty, so nothing

File: packages/migrate/src/__tests__/MigrateDiff.test.ts (hunk 3 above)
Comment (Jolg42) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:94]: What do you mean?

Note for myself: the error message is confusing here, to check and fix

File: packages/migrate/src/__tests__/MigrateDiff.test.ts (hunk 4 above)
Comment (millsp) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:122]: Ah sorry, I meant https://github.com/prisma/prisma/pull/11440/files/c4d0402834ee1603f39a05bcb74310edaf569539#diff-06b471162d0b222816569f260465ceefd7e0a609009a4a5c81ca8e83cb218db7R135

Comment (millsp) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:122]: But there should be an error here on this at least?

File: packages/migrate/src/__tests__/MigrateDiff.test.ts (hunk 3 above)
Comment (millsp) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:94]: To have a 'prisma.schema' in the package.json

Comment (Jolg42) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:94]: Yes, for example, generate calls `getSchemaPathInternal` which 
```
  // 1. Try the user custom path
  // 2. Try the package.json `prisma.schema` custom path
  // 3. Try the conventional `./schema.prisma` or `./prisma/schema.prisma` paths
  // 4. Try resolving yarn workspaces and looking for a schema.prisma file there
```
See https://github.com/prisma/prisma/blob/main/packages/sdk/src/cli/getSchema.ts#L284-L295

File: packages/migrate/src/__tests__/MigrateDiff.test.ts [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+    })
+    it('should diff --from-url=file:doesnotexists.db --to-empty ', async () => {
+      ctx.fixture('schema-only-sqlite')
```
Comment (Jolg42) [inline packages/migrate/src/__tests__/MigrateDiff.test.ts:125]:
```suggestion
      ctx.fixture('introspection/sqlite')
```

File: packages/migrate/src/__tests__/MigrateDiff.test.ts [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+
+      const result = MigrateDiff.new().parse(['--preview-feature', '--from-empty', '--to-url=file:doesnotexists.db'])
+      await expect(result).resolves.toMatchInlineSnapshot(``)
+      expect(ctx.mocked['console.info'].mock.calls.join('\n')).toMatchInlineSnapshot(``)
+    })
+
+    it('should diff --from-empty --to-url=file:dev.db', async () => {
+      ctx.fixture('schema-only-sqlite')
```
Comment (Jolg42) [inline packages/migrate/src/__tests__/MigrateDiff.test.ts:133]:
```suggestion
      ctx.fixture('introspection/sqlite')
```

File: packages/migrate/src/__tests__/MigrateDiff.test.ts (hunk 3 above)
Comment (millsp) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:94]: TIL :100:

File: packages/migrate/src/__tests__/MigrateDiff.test.ts (hunk 4 above)
Comment (Jolg42) [reply packages/migrate/src/__tests__/MigrateDiff.test.ts:122]: There yes, I'll check what's going on

File: packages/migrate/src/commands/MigrateDiff.ts [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+const debug = Debug('prisma:migrate:diff')
+export class MigrateDiff implements Command {
+  public static new(): MigrateDiff {
+    return new MigrateDiff()
+  }
+
+  private static help = format(`
+${
+  process.platform === 'win32' ? '' : chalk.bold('🔍 ')
+}Compares the database schema from two arbitrary sources, and displays the differences either as a human-readable summary (by default) or an executable script.
```
Comment (floelhoeffel) [inline packages/migrate/src/commands/MigrateDiff.ts:19]: Should we list here already what those supported sources are?

I'd also change "displays" to "outputs"

Comment (Jolg42) [reply packages/migrate/src/commands/MigrateDiff.ts:19]: >Should we list here already what those supported sources are?
Too long for a one liner here 😅 Better to keep it short?

**Pull request merged**